optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,256 @@
+from collections import defaultdict
+import itertools
+from typing import Any
+from typing import DefaultDict
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import numpy as np
+
+import optuna
+from optuna._experimental import experimental
+from optuna.distributions import BaseDistribution
+from optuna import multi_objective
+from optuna.multi_objective.samplers import BaseMultiObjectiveSampler
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "multi_objective:nsga2:generation"
+_PARENTS_KEY = "multi_objective:nsga2:parents"
+
+
+@experimental("1.4.0")
+class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
+    """Multi-objective sampler using NSGA-II algorithm.
+
+    NSGA-II stands for "Nondominated Sorting Genetic Algorithm II",
+    which is a well known, fast and elitist multi-objective genetic algorithm.
+
+    For further information about NSGA-II, please refer to the following paper:
+
+    - `A fast and elitist multiobjective genetic algorithm: NSGA-II
+      <https://ieeexplore.ieee.org/document/996017>`_
+
+    Args:
+        seed:
+            Seed for random number generator.
+
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+
+        crossover_prob:
+            Probability of swapping each parameter of the parents when creating a new individual.
```

Comment:
I agree with @hvy.
I thought this probability was to determine whether two chromosomes are recombined or not.

> (which I'd suspect usually is 0.5, equal probability for both parents which also makes me wonder whether this arguments is really needed since the parents are chosen randomly uniformly).

I agree. [This article](https://pdfs.semanticscholar.org/8ed1/d9f5a017c61abfec4089aba828b398aa3e7e.pdf) also explains that this probability is usually 0.5 (see `swapping_probability` in the description of uniform crossover.)

I personally prefer adding `crossover_prob` in order to specify the probability of whether a crossover should occur, and fixing the `swapping_probability` to `0.5`. In this case, the value of `crossover_prob` seems to be 0.9 in Deb's paper.

Reply:
@hvy @toshihikoyanase Thanks! You're definitely right. I renamed the existing `crossover_prob` with `swapping_prob` and added the correct `crossover_prob` parameter. The default value of each parameter are also fixed (https://github.com/optuna/optuna/pull/1163/commits/fbaf12e03546c47d6527b1574dd3a19c8c2d777d).

Reply:
Besides, I modified the type of `mutation_prob` to allow taking `None` at the commit https://github.com/optuna/optuna/pull/1163/commits/89457170d9942289d1ead5f2600c4f0493e4d6cf. If `None` is passed,  `1 / number of parent trial parameters` is used as the value. This behavior seems to align to  Deb's paper.